astronomy/solar_system_dynamics_test.cpp

```diff
@@ -0,0 +1,352 @@
+      expected_subsystem_barycentre.Add(
+          expected_system.initial_state(moon_name),
+          moon->gravitational_parameter());
+    }
+
+    auto const actual_dof = actual_subsystem_barycentre.Get();
+    auto const expected_dof = expected_subsystem_barycentre.Get();
+
+    auto const actual_parent_dof =
+        ephemeris.trajectory(parent)->
+            EvaluateDegreesOfFreedom(epoch, /*hint=*/nullptr);
+    auto const expected_parent_dof =
+        expected_system.initial_state(parent_name);
+
+    // We transform to a frame in which |parent| has the z-axis as its rotation
+    // axis by rotating around the normal to Earth's and |parent|'s rotation
+    // axes.
+    // If |parent| is the Sun, we use the normal to the invariable plane instead
+    // of the Sun's axis.
+    // TODO(egg): perhaps rotating bodies should export a rotation to their
+    // celestial reference frame, we'll use that in the plugin too.
+    enum LocalFrameTag { tag };
+    using ParentEquator = Frame<LocalFrameTag, tag, /*frame_is_inertial=*/true>;
+    auto const z = Bivector<double, ICRFJ2000Equator>({0, 0, 1});
+    std::experimental::optional<Rotation<ICRFJ2000Equator, ParentEquator>>
+        rotation;
+
+    if (SolarSystemFactory::parent(index) == SolarSystemFactory::Sun) {
+      Bivector<AngularMomentum, ICRFJ2000Equator>
+          solar_system_angular_momentum;
+      for (int i = 1; i <= SolarSystemFactory::LastBody; ++i) {
+        auto const body_name = SolarSystemFactory::name(i);
+        auto const body = system.massive_body(ephemeris, body_name);
+        RelativeDegreesOfFreedom<ICRFJ2000Equator> const
+            from_solar_system_barycentre =
+                expected_system.initial_state(body_name) -
+                DegreesOfFreedom<ICRFJ2000Equator>(ICRFJ2000Equator::origin,
+                                                   {});
+        solar_system_angular_momentum +=
+            Wedge(from_solar_system_barycentre.displacement(),
+                  body->mass() * from_solar_system_barycentre.velocity()) *
+            Radian;
+      }
+      auto const normal =
+          Commutator(Normalize(solar_system_angular_momentum), z);
+      // Check that we computed the invariable plane properly by computing its
+      // angle to the Sun's equator.
+      // The actual figure is "almost exactly 6 deg", see Bailey, Batygin, and
+      // Brown, Solar Obliquity Induced by Planet Nine,
+      // https://arxiv.org/pdf/1607.03963.pdf.
+      EXPECT_THAT(AngleBetween(solar_system_angular_momentum,
+                               parent->angular_velocity()),
+                  AllOf(Gt(5.9 * Degree), Lt(6.0 * Degree)));
+      auto const declination_of_invariable_plane =
+          AngleBetween(solar_system_angular_momentum, z);
+      rotation =
+          Rotation<ICRFJ2000Equator, ParentEquator>(
+              declination_of_invariable_plane,
+              normal);
+    } else {
+      auto const ω = parent->angular_velocity();
```

Comment:
In general this code seems to overuse `auto` a bit.  I would rather see a real type in many of these declarations.